source/mge.js

```diff
@@ -0,0 +1,988 @@
+                        break;
+                    default:
+                        throw new megengine.Error("Unsupported tensor data type '" + context.dataType + "'.");
+                }
+            }
+        }
+        else {
+            for (let j = 0; j < size; j++) {
+                if (context.count > context.limit) {
+                    results.push('...');
+                    return results;
+                }
+                results.push(this._decode(context, dimension + 1));
+            }
+        }
+        if (context.dimensions.length == 0) {
+            return results[0];
+        }
+        return results;
+    }
+
+    static _stringify(value, indentation, indent) {
+        if (Array.isArray(value)) {
+            const result = [];
+            result.push(indentation + '[');
+            const items = value.map((item) => megengine.Tensor._stringify(item, indentation + indent, indent));
+            if (items.length > 0) {
+                result.push(items.join(',\n'));
+            }
+            result.push(indentation + ']');
+            return result.join('\n');
+        }
+        if (value && (value instanceof base.Int64 || value instanceof base.Uint64)) {
+            return indentation + value.toString();
+        }
+        if (typeof value == 'string') {
+            return indentation + value;
+        }
+        if (value == Infinity) {
+            return indentation + 'Infinity';
+        }
+        if (value == -Infinity) {
+            return indentation + '-Infinity';
+        }
+        if (isNaN(value)) {
+            return indentation + 'NaN';
+        }
+        return indentation + value.toString();
+    }
+};
+
+megengine.TensorType = class {
+
+    constructor(dataType, shape) {
+        this._dataType = dataType;
+        this._shape = shape;
+    }
+
+    get dataType() {
+        return this._dataType;
+    }
+
+    get shape() {
+        return this._shape;
+    }
+
+    toString() {
+        return this._dataType + this._shape.toString();
+    }
+};
+
+megengine.TensorShape = class {
+
+    constructor(dimensions) {
+        this._dimensions = dimensions || [];
+    }
+
+    get dimensions() {
+        return this._dimensions;
+    }
+
+    toString() {
+        if (this._dimensions && this._dimensions.length > 0) {
+            return '[' + this._dimensions.map((dimension) => dimension.toString()).join(',') + ']';
+        }
+        return '';
+    }
+};
+
+megengine.Execution = class extends python.Execution {
+
+    constructor(sources, exceptionCallback) {
+        super(sources, exceptionCallback);
+        this.registerType('megengine.tensor.Parameter', class {
+            constructor(data, dtype, device) {
+                this.data = data;
+                this.dtype = dtype;
+                this.device = device;
+            }
+        });
+        this.registerType('megengine.traced_module.pytree.TreeDef', class {
+            toString() {
```

Comment:
These methods do not exist in the Python implementation. Formatting should be done in client code, `registerType` should only be used to match Python execution and serialization.

Reply:
`toString` method is needed in serialization, the treedef object may be a key of `argdef_graph_map`, without it the dict will be loaded as `[[object]] : xxxx`.

`formatString` method has been moved to `megengine.Utility` as a static method.